panther_gpiod/include/panther_gpiod/gpio_driver.hpp

```diff
@@ -0,0 +1,291 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file gpio_driver.hpp
+ *
+ * @brief Header file containing declarations for the GPIODriver class
+ *        responsible for GPIO operations on Panther robot.
+ *
+ * This file contains the declarations for the GPIODriver class, which
+ * manages GPIO pins on the Panther robot by providing functionalities to set pin values,
+ * change pin directions, monitor pin events, and more.
+ */
+
+#ifndef PANTHER_GPIOD_GPIO_DRIVER_HPP_
+#define PANTHER_GPIOD_GPIO_DRIVER_HPP_
+
+#include <atomic>
+#include <condition_variable>
+#include <filesystem>
+#include <functional>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <shared_mutex>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include <gpiod.hpp>
+
+namespace panther_gpiod
+{
+
+/**
+ * @brief Enumeration representing available GPIO pins in the Panther system.
+ */
+enum class GPIOPin {
+  AUX_PW_EN,
+  CHRG_DISABLE,
+  CHRG_SENSE,
+  DRIVER_EN,
+  E_STOP_RESET,
+  FAN_SW,
+  GPOUT1,
+  GPOUT2,
+  GPIN1,
+  GPIN2,
+  LED_SBC_SEL,
+  SHDN_INIT,
+  STAGE2_INPUT,
+  VDIG_OFF,
+  VMOT_ON,
+  WATCHDOG
+};
+
+/**
+ * @brief Structure containing information related to GPIO pins such as pin configuration,
+ * direction, value, etc. This information is required during the initialization process.
+ */
+struct GPIOInfo
+{
+  GPIOPin pin;
+  gpiod::line::direction direction;
+  bool active_low = false;
+  gpiod::line::value init_value = gpiod::line::value::INACTIVE;
+  gpiod::line::value value = gpiod::line::value::INACTIVE;
+  gpiod::line::offset offset = -1;
+};
+
+/**
+ * @brief Class responsible for managing GPIO pins on Panther robots, handling tasks such as
+ * setting pin values, changing pin directions, monitoring pin events, and more.
+ */
+class GPIODriver
+{
+public:
+  /**
+   * @brief Constructs the GPIODriver object with information about GPIO pin configurations.
+   * This information is necessary for initializing the GPIO functionality.
+   *
+   * @param gpio_info_storage Vector containing information about GPIO pin configurations.
+   *
+   * @throws std::runtime_error if the provided `gpio_info_storage` vector is empty.
+   *
+   * @par Example
+   * An example of constructing the GPIODriver object by providing GPIO pin information:
+   * @code{.cpp}
+   * std::vector<GPIOInfo> gpio_configurations = {
+   *   GPIOInfo{GPIOPin::CHRG_SENSE, gpiod::line::direction::INPUT},
+   *   GPIOInfo{GPIOPin::AUX_PW_EN, gpiod::line::direction::OUTPUT},
+   *   GPIOInfo{GPIOPin::LED_SBC_SEL, gpiod::line::direction::OUTPUT, true,
+   * gpiod::line::value::ACTIVE}
+   *   // ... additional GPIO pin configurations
+   * };
+   * GPIODriver gpio_driver(gpio_configurations);
+   * @endcode
+   */
+  GPIODriver(std::vector<GPIOInfo> gpio_info_storage);
```

Comment:
I would also add info that to enable reading pin values it is required to enable gpio monitoring, see GPIOMonitorEnable method for more info.